csi.proto

```diff
@@ -502,6 +487,176 @@ message ControllerServiceCapability {
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 4;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // SnapshotSize is the disk size of the snapshot. The purpose of this
+  // field is to give CO guidance on how much space is needed to create
+  // a volume from this snapshot. The size of the volume must not be
+  // less than the size of the source snapshot. This field is OPTIONAL.
+  // If SnapshotSize is not set, it indicates that this size is unknown.
+  SnapshotSize size = 1;
+
+  // Contains identity information for the created snapshot. This field
+  // is REQUIRED. The identity information will be used by the CO in
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. This field is REQUIRED.
+  string timestamp = 4;
+
+  // The status of a snapshot. A snapshot could have the following
+  // statuses:
+  // UPLOADING - A snapshot is cut and is now being uploaded.
+  // READY - A snapshot is ready for use.
+  // Some cloud providers will upload the snapshot to a location in
+  // the cloud (i.e., an object store) after the snapshot is cut.
+  // Uploading is a long process that could take hours. At this phase,
+  // since the snapshot is already cut, a `thaw` operation can be
+  // performed so application can start running again (assuming a
+  // `freeze` operation was done before taking the snapshot). The
+  // status of the snapshot will become `READY` after the upload is
+  // complete.
+  // For cloud providers and storage systems that don't have the
+  // uploading process, the status should be `READY` after the snapshot
+  // is cut. `thaw` can be done when the status is `READY` in this case.
+  // A `CREATING` status is not included here because CreateSnapshot
+  // is synchronous and will block until the snapshot is cut.
+  // An `ERROR` status is also not included because there should be an
+  // gRPC error code if an error happens.
+  SnapshotStatus status = 5;
+}
+
+message SnapshotSize {
+  // The disk size of the snapshot. The purpose of this field is to
+  // give CO guidance on how much space is needed to create a volume
+  // from this snapshot. The size of the volume must not be less than
+  // the size of the source snapshot. This field is OPTIONAL. The value
+  // of this field MUST NOT be negative.
+  int64 size = 1;
```

Comment:
Maybe `size_bytes`? Does this need it's own message spec?

Reply:
Sure, I'll rename this to size_bytes.  The reason we use a message spec is because some storage systems do not know the size of volume or snapshot and therefore will leave this field as null.  Using a message spec makes it easier to differentiate between null and a size of 0.

Reply:
Changed to size_bytes.